### PR TITLE
[PTO] Add T.tile.row_expand_mul API and fix DMA stride / transpose codegen issues

### PR DESCRIPTION
**Pull request by handsomeRobotSK**

### CODE REVIEW

**Comment by handsomeRobotSK** [comment]
## Commit 1：`[PTO] Add T.tile.row_expand_mul API`

### 1.1 背景

Block Sparse MQA Attention kernel 需要对 score 矩阵做 `relu(scores) * weights`，其中 `weights` 是 `[heads]` 的行向量。PTO ISA 提供了 `TROWEXPANDMUL` 硬件指令：`dst[i,j] = src0[i,j] * src1[i,0]`，将一个行向量广播到矩阵所有列再做逐元素乘法。此前 tilelang 没有封装这个指令。

### 1.2 Python API：`row_expand_mul()`（`tilelang/language/ascend_tile.py`）

**做了什么**：新增 `T.tile.row_expand_mul(dst, src0, src1, tmp=None)` 函数。

**参数约定**：
- `dst`/`src0`：2D buffer 或 BufferRegion，shape `[rows, cols]`
- `src1`：行向量，支持 1D `[R]`、2D `[1,R]`、2D `[R,1]` 三种形态
- `tmp`：可选临时 buffer

**为什么这样设计**：
- 支持 BufferRegion 切片参数，因为 kernel 中 buffer 常被切片（如 `s_ub[:, 0*kv:1*kv]`）
- 自动展开 src1 的三种形状，降低用户心智负担
- 通过 `tir.call_intrin` 调用 `tl.ascend_row_expand_mul` op，走标准 tilelang intrinsic 路径

**新增 `_handle_buffer_region_2d()` 辅助函数**：
- `_handle_buffer_region()` 返回 ND extent，但 `GetSliceInfo` 只处理 2D。对于 3D buffer（如 `s_ub[T, 32, 512]`）切片 `s_ub[tile, :, :]`，原始 `_handle_buffer_region` 返回 `[1, 32, 512]` 三维 extent
- `_handle_buffer_region_2d` 将 ND extent 折叠为 `[rows, cols]`：fold 所有前导维度到 rows，保留最后一维为 cols。例如 `[1, 32, 512]` → `[32, 512]`

### 1.3 Op 注册（`src/op/ascend.cc` / `ascend.h`）

**做了什么**：注册 `ascend_row_expand_mul` op，`set_num_inputs(-1)` 表示可变参数（tmp 可选）。

**为什么用 -1**：带 tmp 是 5 个参数，不带是 4 个。`-1` 让 TVM 不校验参数数量。

### 1.4 Codegen 分发（`src/target/codegen_ascend.cc` / `.h`）

**做了什么**：
- `VisitExpr_` 中添加 `ascend_row_expand_mul` → `RowExpandMulCodegen(op)` 分发
- 基类 `CodeGenTileLangAscend` 中添加 `LOG(FATAL)` stub

**为什么非 PTO 路径要 stub**：`TROWEXPANDMUL` 是 PTO ISA 特有指令。`CodeGenTileLangAscend` 是非 PTO 的 AscendC 路径，不会走到这里，但编译期需要方法定义。如果错误走到此路径，`LOG(FATAL)` 给出明确报错而不是链接错误。

### 1.5 PTO Codegen 实现（`src/target/codegen_ascend_pto.cc` / `.h`）

**做了什么**：`RowExpandMulCodegen()` 方法，生成如下代码：
```cpp
tl::ascend_pto::TROWEXPANDMUL_row_vec<float, 32, 512, 32>(
    s_ub_4x, s_ub_4x, weights, 65600, 0);
```

**关键设计 —— `fix_nd_2d` lambda**：

`GetSliceInfo` 只能处理最多 2D 的 buffer shape。对于 3D+ buffer（如 `s_ub[T, 32, 512]`），它会丢弃尾随维度，导致 `slice_row`/`slice_col` 互换。`fix_nd_2d` 从 `buffer_shapess_` 获取原始 buffer shape，重新推导正确的：
- `slice_col` = 最内层维度
- `slice_row` = extent / 最内层维度
- `slice_valid_col` = 最内层维度
- `slice_valid_row` = slice_row

**关键设计 —— `fix_src1_nd` lambda**：

src1 是一个 1D 行向量（如 `weights[H]`），但 may be stored in a 2D+ buffer（如 `weights_ub[1, H]`）。当 src1 被切片时，`GetSliceInfo` 可能产生错误的 `slice_valid_col`。`fix_src1_nd` 从原始 buffer shape 提取正确的最后维度值。

**为什么需要临时变量（`CreateUbVariableND`）**：

当 buffer 被切片时，直接传入切片区域会导致模板参数和实际 buffer 维度不匹配。$codegen 先创建一个全尺寸的临时 `TileUbDataND`，再将切片数据映射进去，最后传入模板。

### 1.6 PTO Template（`src/tl_templates/pto/common.h`）

**做了什么**：新增 `TROWEXPANDMUL_row_vec` 模板函数（两个重载）。

**为什么要把 RowMajor 行向量转成 ColMajor**：

src1 在 UB 中按 RowMajor 存储为 `[1, vecLen]`。`TROWEXPANDMUL` 要求 src1 是 ColMajor 列向量 `[vecLen, 1]`。由于 RowMajor `[1, N]` 和 ColMajor `[N, 1]` 在内存中物理布局相同，可以通过创建 `TileUbDataDN`（ColMajor tile）并 `TASSIGN` 到同一地址，零拷贝"重新解释"内存布局。

**为什么用 32B 对齐**：PTO ISA 要求 tile 地址 32B 对齐。`alignedRows = ((vecLen * sizeof(T) + 31) / 32) * (32 / sizeof(T))` 计算对齐后的行数。

---

## Commit 2：`[Fix] DMA stride mismatch, PadValue, L1→L0 transpose`

### 2.1 Column-Slice DMA Stride 修复（`GMCopyCall`）

**问题**：

当 `T.copy(src, ub_buf[:, off:off+kv])` 写入 UB buffer 的列切片时，生成的 DMA tile 使用了切片的维度而非 buffer 的完整维度。

具体来说，buffer `s_ub[32, 512]`，切片 `s_ub[:, 0:128]`：
- 旧代码：`shape_tmpl = [1, 1, 1, 32, 128]`，tile = `TileUbDataND<float, 32, 128>`
- `TileUbDataND` 的 row stride = `Cols * sizeof(T)` = `128 * 4 = 512` bytes
- 但 buffer `s_ub` 的真实 row stride = `512 * 4 = 2048` bytes
- **结果**：第 1 行之后的数据写入错误的字节偏移，破坏相邻列数据

**修复**：

```cpp
// 旧：
shape4 = slice_info.is_slice ? slice_info.slice_valid_row : slice_info.row;
shape5 = slice_info.is_slice ? slice_info.slice_valid_col : slice_info.col;
// 新：
shape4 = slice_info.row;  // buffer 完整行数
shape5 = slice_info.col;  // buffer 完整列数
```

tile 参数也使用完整 `row`/`col`，确保 stride 匹配。`valid_row`/`valid_col` 仍然控制实际传输的元素数。

**不改的后果**：列切片 DMA 在多行场景下数据写入错误位置，计算静默出错误结果。

### 2.2 PadValue::Null 修复

**问题**：

DMA stride 修复后，UB tile 使用 buffer 完整列数（如 512），但 valid_col 只有切片大小（如 128）。PTO 编译器 pass `TFILLPAD_INPLACE` 检测到 `valid_col < ub_shape2`，自动在 DMA 后对非 valid 区域（列 128~511）填充零。

对于列切片 DMA，这会把零写入到同 buffer 中**其他切片正在使用的列**，造成数据破坏。

**修复**：

```cpp
if (slice_info.is_slice) {
    stream << "pto::PadValue::Null";  // 禁用 fillpad
} else {
    stream << GetPadEnum(call->args[6]);  // 用户指定的 pad 策略
}
```

`PadValue::Null` 告诉编译器跳过 fillpad。因为切片场景下，valid region 就是完整需要的数据，没有 tail padding 的需求。

### 2.3 L1→L0 Transpose Tile Offset 修复（`CopyL1ToL0Codegen`）

**问题**：

`T.copy(k_l1, l0b, transpose=True)` 在 L1→L0B 带转置的场景下：

Bug 1 — TileMatL1ZN 模板参数：
```cpp
// 旧：
TileMatL1ZN<T, slice_col, slice_row, slice_col, slice_row>
// 一个 tile 只需要 tile_row × tile_col，不是整个 slice
// 新：
TileMatL1ZN<T, tile_col, tile_row, tile_col, tile_row>
```

Bug 2 — TASSIGN 地址偏移：
```cpp
// 旧：
TASSIGN(zn, base_addr + src_shape_info.offset * typeLen)
// offset 是第一块 tile 内的偏移，不随 tile 索引变化
// 新：
TASSIGN(zn, base_addr + outer_tile_idx * tile_size * typeLen)
```

当逻辑 slice 大小等于一个 tile 时，两个 bug 恰好不暴露（`slice_row == tile_row`，`offset == tile_base_offset`）。但当 slice > tile（需要多 tile 循环时），第二块及之后的 tile 会引用错误的内存地址。

**不改的后果**：多 tile 循环场景下，第 2+ 个 tile 的源数据地址错误，计算结果静默出错。

---

## Commit 3：Examples

两个完整的 Block Sparse MQA Attention kernel，用作功能验证和性能测试。实现了手动 CV scope、手动 sync/flag、显式 L0A/L0B/MMA 的 expert 模式。

---

## 常见检视问题

**Q1: 为什么不直接用 `_handle_buffer_region` 而要加 `_handle_buffer_region_2d`？**

> `_handle_buffer_region` 返回 ND extent（如 `[1, 32, 512]`），但 `row_expand_mul` 内部逻辑和下游 `GetSliceInfo` 都需要 2D `[rows, cols]`。3D extent 传下去会导致 rows/cols 互换的 bug。`_handle_buffer_region_2d` 在前端就把问题解决。

**Q2: `fix_nd_2d` / `fix_src1_nd` 和 `_handle_buffer_region_2d` 不重复吗？**

> 不重复。Python 端的 `_handle_buffer_region_2d` 修正了 extent 传给 access_ptr 的方式，但它不影响 C++ codegen 端 `GetSliceInfo` 对 buffer 原始 shape 的解析。`fix_nd_2d` 是 C++ 端的二次保险：当 `GetSliceInfo` 仍因 3D+ buffer 产生错误 slice 信息时，从 `buffer_shapess_` 重新推导正确值。

**Q3: 为什么 GMCopyCall 要用 buffer 完整维度而不是 slice 维度？**

> `TileUbDataND` 的物理行 stride = `Cols * sizeof(T)`。如果 UB tile 用 slice 维度（如 128 列），stride 就是 128×4=512 bytes。但目标 buffer `s_ub[32, 512]` 的真实行 stride 是 512×4=2048 bytes。DMA 写到 tile 第 2 行时按 512 bytes stride 写，实际写到了 buffer 的第 512~640 字节位置（应该是第 2048~2560 字节），数据错位。

**Q4: 为什么 PadValue::Null 可以安全地替代 PadValue::Zero？**

> 列切片场景下，我们只关心 `[0:128]` 列的数据。这些列在 DMA 后已被完整填充，不需要 tail padding。`PadValue::Null` 只是跳过 fillpad pass，不影响实际传输的正确数据。真正需要 zero-padding 的场景（如处理非对齐的矩阵尾块）仍然使用 `PadValue::Zero`。

**Q5: TileMatL1ZN 的 bug 为什么之前没被发现？**

> 两个 bug 在 `slice size == tile size` 时恰好抵消（`slice_row == tile_row`，`offset == tile_base_offset`）。之前的用例可能都小于等于一个 tile。Block Sparse MQA kernel 中 L1 buffer 比 tile 大，才触发了这个 bug。

**Review by fuhouyu-hw — DISMISSED** [review]
/lgtm

File: examples/HISA/block_sparse_mqa_attn_expert_test.py [hunk 1]
```diff
@@ -0,0 +1,690 @@
...
+import tilelang
+from tilelang import language as T
+import argparse
+import torch
+
+tilelang.disable_cache()
+
+
+@tilelang.jit(
+    pass_configs={
```
Comment (LLMZhangYC) [inline examples/HISA/block_sparse_mqa_attn_expert_test.py]: Delete when using no developer feature.

File: examples/HISA/block_sparse_mqa_attn_expert_test.py [hunk 2]
```diff
@@ -0,0 +1,690 @@
+"""
+该脚本用于在expert模式下直接开始修改
```
Comment (LLMZhangYC) [inline examples/HISA/block_sparse_mqa_attn_expert_test.py]: Comments should in English. The same problem in next python file.

**Review by LLMZhangYC — APPROVED** [review]
lgtm

File: examples/HISA/block_sparse_mqa_attn_expert_test.py (hunk 2 above)
Comment (handsomeRobotSK) [reply examples/HISA/block_sparse_mqa_attn_expert_test.py]: OK

File: examples/HISA/block_sparse_mqa_attn_expert_test.py (hunk 1 above)
Comment (handsomeRobotSK) [reply examples/HISA/block_sparse_mqa_attn_expert_test.py]: Copy that

**Review by fuhouyu-hw — APPROVED** [review]
/approve
